test/e2e/helpers_test.go

```diff
@@ -0,0 +1,277 @@
+	}
+}
+
+func cleanupEtcdCluster(ctx context.Context, t *testing.T, c *envconf.Config, name string) {
+	t.Helper()
+	var etcdCluster ecv1alpha1.EtcdCluster
+	if err := c.Client().Resources().Get(ctx, name, namespace, &etcdCluster); err == nil {
+		if err := c.Client().Resources().Delete(ctx, &etcdCluster); err != nil {
+			t.Logf("Failed to delete EtcdCluster: %v", err)
+		}
+	}
+}
+
+// getEtcdMembersName2IDMapping retrieves the etcd cluster member list as name->ID mapping using etcd's native types
+func getEtcdMembersName2IDMapping(t *testing.T, c *envconf.Config, podName string) map[string]uint64 {
+	t.Helper()
+	memberList := getEtcdMemberListPB(t, c, podName)
+
+	// Create name->ID mapping
+	memberMap := make(map[string]uint64)
+	for _, member := range memberList.Members {
+		memberMap[member.Name] = member.ID
+	}
+	return memberMap
+}
+
+// getEtcdMemberListPB returns the etcdserverpb.MemberListResponse by calling etcdctl -w json.
+func getEtcdMemberListPB(t *testing.T, c *envconf.Config, podName string) *etcdserverpb.MemberListResponse {
+	t.Helper()
+	stdout, stderr, err := execInPod(t, c, podName, namespace, []string{"etcdctl", "member", "list", "-w", "json"})
+	if err != nil {
+		t.Fatalf("Failed to get etcd member list: %v, stderr: %s", err, stderr)
+	}
+	var memberList etcdserverpb.MemberListResponse
+	if err := json.Unmarshal([]byte(stdout), &memberList); err != nil {
+		t.Fatalf("Failed to parse etcd member list JSON: %v", err)
+	}
+	return &memberList
+}
+
+// waitForNoLearners waits until the member list has the expected number of members
+// and all members are voting (i.e., no learners remain).
+func waitForNoLearners(t *testing.T, c *envconf.Config, podName string, expectedMembers int) {
+	t.Helper()
+	err := wait.For(func(ctx context.Context) (bool, error) {
+		ml := getEtcdMemberListPB(t, c, podName)
+		if len(ml.Members) != expectedMembers {
+			return false, nil
+		}
+		for _, m := range ml.Members {
+			if m.IsLearner {
+				return false, nil
+			}
+		}
+		return true, nil
+	}, wait.WithTimeout(3*time.Minute), wait.WithInterval(5*time.Second))
+	if err != nil {
+		t.Fatalf("Timeout waiting for %d voting members with no learners: %v", expectedMembers, err)
+	}
+}
+
+// verifyPodUsesPVC checks that a pod is using PVC for persistent storage
+func verifyPodUsesPVC(t *testing.T, c *envconf.Config, podName string, expectedPVCPrefix string) {
+	t.Helper()
+	var pod corev1.Pod
+	if err := c.Client().Resources().Get(t.Context(), podName, namespace, &pod); err != nil {
+		t.Fatalf("Failed to get pod %s: %v", podName, err)
+	}
+
+	// Check for PVC volumes
+	hasPVC := false
+	for _, volume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim != nil {
+			if strings.HasPrefix(volume.PersistentVolumeClaim.ClaimName, expectedPVCPrefix) {
+				hasPVC = true
+				break
+			}
+		}
+	}
+
+	if !hasPVC {
+		t.Errorf("Pod %s does not use expected PVC with prefix %s", podName, expectedPVCPrefix)
+	}
```

Comment:
I think this can be further simplified to:

```suggestion
	for _, volume := range pod.Spec.Volumes {
		if volume.PersistentVolumeClaim != nil {
			if strings.HasPrefix(volume.PersistentVolumeClaim.ClaimName, expectedPVCPrefix) {
				return
			}
		}
	}

	t.Errorf("Pod %s does not use expected PVC with prefix %s", podName, expectedPVCPrefix)
```

Reply:
Applied, it's cleaner now.